test/integration/src/test/java/io/pravega/test/integration/ControllerMetricsTest.java

```diff
@@ -0,0 +1,179 @@
+import lombok.Cleanup;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.curator.test.TestingServer;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import static io.pravega.shared.MetricsNames.ABORT_TRANSACTION;
+import static io.pravega.shared.MetricsNames.COMMIT_TRANSACTION;
+import static io.pravega.shared.MetricsNames.CREATE_TRANSACTION;
+
+/**
+ * Check the end to end correctness of metrics published by the Controller.
+ */
+@Slf4j
+public class ControllerMetricsTest {
+
+    private final int controllerPort = TestUtils.getAvailableListenPort();
+    private final String serviceHost = "localhost";
+    private final URI controllerURI = URI.create("tcp://" + serviceHost + ":" + controllerPort);
+    private final int servicePort = TestUtils.getAvailableListenPort();
+    private final int containerCount = 4;
+    private TestingServer zkTestServer;
+    private PravegaConnectionListener server;
+    private ControllerWrapper controllerWrapper;
+    private ServiceBuilder serviceBuilder;
+    private ScheduledExecutorService executor;
+    private StatsProvider statsProvider = null;
+
+    @Before
+    public void setUp() throws Exception {
+        MetricsConfig metricsConfig = MetricsConfig.builder()
+                                                   .with(MetricsConfig.ENABLE_CSV_REPORTER, false)
+                                                   .with(MetricsConfig.ENABLE_STATSD_REPORTER, false)
+                                                   .build();
+        metricsConfig.setDynamicCacheEvictionDurationMs(300000);
+
+        MetricsProvider.initialize(metricsConfig);
+        statsProvider = MetricsProvider.getMetricsProvider();
+        statsProvider.start();
+        log.info("Metrics Stats provider is started");
+
+        executor = Executors.newSingleThreadScheduledExecutor();
+        zkTestServer = new TestingServerStarter().start();
+
+        serviceBuilder = ServiceBuilder.newInMemoryBuilder(ServiceBuilderConfig.getDefaultConfig());
+        serviceBuilder.initialize();
+        StreamSegmentStore store = serviceBuilder.createStreamSegmentService();
+
+        server = new PravegaConnectionListener(false, servicePort, store);
+        server.startListening();
+
+        controllerWrapper = new ControllerWrapper(zkTestServer.getConnectString(),
+                false,
+                controllerPort,
+                serviceHost,
+                servicePort,
+                containerCount);
+        controllerWrapper.awaitRunning();
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        if (this.statsProvider != null) {
+            statsProvider.close();
+            statsProvider = null;
+            log.info("Metrics statsProvider is now closed.");
+        }
+
+        ExecutorServiceHelpers.shutdown(executor);
+        controllerWrapper.close();
+        server.close();
+        serviceBuilder.close();
+        zkTestServer.close();
```

Comment:
Suggest use try-with-resource to ensure clean closure of resource. We often run into time-out problems during unit tests, some of them are exactly due to resources are not released completed after each test, particularly when multiple resources involved.

Here when an proceeding exception is thrown, the following close() method will be skipped.

Seems TestingServer, ControllerWrapper, PravegaConnectionListener, ServiceBuilder are all AutoCloseable, so we could make a good use of them.

Reply:
@RaulGracia cool, makes sense.